File: Java/SlidingWindow.java
/**
 * https://leetcode.com/problems/minimum-window-substring/
 */
public class SlidingWindow {

    public static void main(String[] args) {
        System.out.println(minWindow("ADOBECODEBANC", "ABC").equals("BANC"));
        System.out.println(minWindow("a", "a").equals("a"));
        System.out.println(minWindow("a", "aa").equals(""));
    }

    public static String minWindow(String s, String t) {
        int[] map = new int[128];
        for (char c : t.toCharArray()) {
            map[c]++;
        }
        int start = 0, end = 0, minStart = 0, minLen = Integer.MAX_VALUE, counter = t.length();
        while (end < s.length()) {
            char c1 = s.charAt(end);
            if (map[c1] > 0)
                counter--;
            map[c1]--;
            end++;
            while (counter == 0) { // valid window, now try to reduce it
                if (minLen > end - start) {
                    minLen = end - start;
                    minStart = start;
                }
                char c2 = s.charAt(start);
                map[c2]++;
                if (map[c2] > 0)
                    counter++;
                start++;
            }
        }

        return minLen == Integer.MAX_VALUE ? "" : s.substring(minStart, minStart + minLen);
    }
}

File: Python/SlidingWindow.py
def minWindow(s: str, t: str) -> str:
    if len(t) > len(s):
        return ""
    map = [0] * 128
    for c in t:
        map[ord(c)] += 1
    start = 0
    end = 0
    minStart = 0
    minLen = float('inf')
    counter = len(t)
    while end < len(s):
        c1 = s[end]
        if map[ord(c1)] > 0:
            counter -= 1
        map[ord(c1)] -= 1
        end += 1
        while counter == 0:
            if minLen > end - start:
                minLen = end - start
                minStart = start
            c2 = s[start]
            map[ord(c2)] += 1
            if map[ord(c2)] > 0:
                counter += 1
            start += 1
    return s[minStart:minStart + minLen]

print(minWindow("ADOBECODEBANC", "ABC") == "BANC")
print(minWindow("a", "a") == "a")
print(minWindow("a", "aa") == "")